### ISSUE TITLE
(Co-located) join pruning performance issues

### ISSUE THREAD

**Issue by anarazel** [citusdata/citus#607]
With
```
DROP TABLE IF EXISTS a;
DROP TABLE IF EXISTS b;

CREATE TABLE a(category_id int NOT NULL, data text NOT NULL);
CREATE TABLE b(category_id int NOT NULL, data text NOT NULL);

SELECT master_create_distributed_table('a', 'category_id', 'hash');
SELECT master_create_distributed_table('b', 'category_id', 'hash');

SELECT master_create_worker_shards('a', 512, 1);
SELECT master_create_worker_shards('b', 512, 1);
```

planning
```
SELECT * FROM a JOIN b USING(category_id);
```

takes
```
Time: 8764.395 ms
```

While we can make actual algorithmic improvements to get away from this being an ```O(n^2)``` (or worse) algorithm, the constants here are quite big.  The biggest issue is the function lookup in ShardIntervalsOverlap(), the second biggest the loading of shard intervals in FragmentInterval().

**Comment by anarazel** [comment]
FWIW, after the fixes of #608, we start to hit performance problems again at 4096 shards upwards.

The non-algorithmic fix would be to pass down the DistTableCacheEntry  in FragementCombinationList(), as we currently spend most of the time doing LookupDistTableCacheEntry()'s. 

An algorithmic fix would be to iterate over sorted lists of both sides of the join, and only compare the ones that actually can match.

**Comment by ozgune** [comment]
@anarazel -- I'm adding a quick comment for documentation purposes. If the following statement doesn't look accurate, could you correct it?

When the user adds a filtering clause on the partition column, and partition pruning executes before the join, we then apply the join pruning algorithm on fewer number of shards. This is why some of our customers who join 100K+ shards didn't observe this issue. When the user's query doesn't have any filters on the partition column, we then run into these performance bottlenecks with join pruning.
